File: src/Solutions/Euler246.hs
{-# LANGUAGE Strict #-}
module Solutions.Euler246 where
import Utils.Operators((<&&>))

{-
Look at associated GeoGebra file.
You can move point on the red circle and see that all interesting points are inside that circle.

https://math.stackexchange.com/questions/1932935/finding-angle-between-two-tangents-to-the-ellipse

Ellipse equation: ((x - 3000) / 7500)^2 + ((y - 1500) / (2500 * sqrt 5))^2 = 1
Equivalently: 5 * (x - 3000)^2 + 9 * (y - 1500)^2 = 5 * 7500^2
Parameterization: (7500 * cos t + 3000, 2500 * sqrt 5 * sin t + 1500)
Suppose that we're investigating point (px, py) outside of ellipse to find tangents.
Then we have: px * cos t / 7500 + py * sin t / (2500 * sqrt 5) = 1
-}

data Circle = Circle { center :: (Int, Int), radius :: Int }
f = fromIntegral :: Int -> Double

dist (a, b) (x, y) = sqrt $ (a - x)^2 + (b - y)^2
minus (a, b) (x, y) = (a - x, b - y)
dot (a, b) (x, y) = a * x + b * y

pointM = (-2000, 1500)
pointG = (8000, 1500)
pointE = (3000, 1500)
circleC = Circle pointM 15000
circleBound = Circle pointE 19000

latticePoints :: Circle -> [(Int, Int)]
latticePoints c = [(x, y) | x <- [cx - r..cx + r],
                            let t = floor $ sqrt $ f $ r^2 - (cx - x)^2,
                            y <- [cy - t..cy + t]] where
    (cx, cy) = center c
    r = radius c

ellipse :: Double -> (Double, Double)
ellipse t = (7500 * cos t + 3000, 2500 * sqrt 5 * sin t + 1500)
outsideEllipse :: (Int, Int) -> Bool
outsideEllipse (x, y) = 5 * (x - 3000)^2 + 9 * (y - 1500)^2 > 5 * 7500^2

-- (7500 * cos t + 3000 - x, 2500 * sqrt 5 * sin t + 1500 - y) x (-7500 * sin t, 2500 * sqrt 5 * cos t)
-- solve (7500 * cos t + 3000 - x) * 2500 * sqrt 5 * cos t + (2500 * sqrt 5 * sin t + 1500 - y) * 7500 * sin t = 0 for t
-- (in WolframAlpha)
properAngle :: (Int, Int) -> Bool
properAngle (x, y) = cosphi < sqrt 0.5 where
    u = sqrt $ f $ 5 * x^2 + 9 * y^2 - 30000 * x - 27000 * y - 216000000
    s = f $ 3 * y - 4500
    d = sqrt 5 * f (x + 4500)
    z1 = ellipse $ if d == 0 then pi else - 2 * atan ((u - s) / d)
    z2 = ellipse $ if d == 0 then 2 * atan (2500 * sqrt 5 / (f y - 1500)) else 2 * atan ((u + s) / d)
    p = (f x, f y)
    cosphi = (dot (minus z1 p) (minus z2 p)) / (dist z1 p * dist z2 p)

properLatticePoints = filter (outsideEllipse <&&> properAngle) (latticePoints circleBound)

euler246 :: IO String
euler246 = return $ show $ length properLatticePoints

File: rust/src/solutions/euler246.rs
use std::f64::consts::PI;

struct Point(i64, i64);

struct Circle {
    center: Point,
    radius: i64,
}

const POINT_E: Point = Point(3000, 1500);

const CIRCLE_BOUND: Circle = Circle {
    center: POINT_E,
    radius: 19000,
};

impl Circle {
    fn lattice_points(&self) -> impl Iterator<Item = Point> {
        let Point(cx, cy) = self.center;
        let r = self.radius;
        (cx - r..=cx + r).flat_map(move |x| {
            let t = ((r * r - (cx - x).pow(2)) as f64).sqrt().floor() as i64;
            (cy - t..=cy + t).map(move |y| Point(x, y))
        })
    }
}

fn dist((a, b): (f64, f64), (x, y): (f64, f64)) -> f64 {
    ((a - x).powi(2) + (b - y).powi(2)).sqrt()
}

fn minus((a, b): (f64, f64), (x, y): (f64, f64)) -> (f64, f64) {
    (a - x, b - y)
}

fn dot((a, b): (f64, f64), (x, y): (f64, f64)) -> f64 {
    a * x + b * y
}

fn ellipse(t: f64) -> (f64, f64) {
    let (sin, cos) = t.sin_cos();
    (7500.0 * cos + 3000.0, 2500.0 * 5f64.sqrt() * sin + 1500.0)
}

impl Point {
    fn outside_of_ellipse(&self) -> bool {
        5 * (self.0 - 3000).pow(2) + 9 * (self.1 - 1500).pow(2) > 5 * 7500 * 7500
    }

    fn proper_angle(&self) -> bool {
        let &Point(x, y) = self;
        let u = ((5 * x * x + 9 * y * y - 30000 * x - 27000 * y - 216000000) as f64).sqrt();
        let s = (3 * y - 4500) as f64;
        let d = 5f64.sqrt() * (x + 4500) as f64;
        let z1 = ellipse(if d == 0.0 {
            PI
        } else {
            -2.0 * ((u - s) / d).atan()
        });
        let z2 = ellipse(if d == 0.0 {
            2.0 * (2500.0 * 5f64.sqrt() / (y - 1500) as f64).atan()
        } else {
            2.0 * ((u + s) / d).atan()
        });
        let p = (x as f64, y as f64);
        let cos_phi = dot(minus(z1, p), minus(z2, p)) / (dist(z1, p) * dist(z2, p));
        cos_phi < 0.5f64.sqrt()
    }
}

pub fn euler246() -> String {
    CIRCLE_BOUND
        .lattice_points()
        .into_iter()
        .filter(Point::outside_of_ellipse)
        .filter(Point::proper_angle)
        .count()
        .to_string()
}
